File: src/ExternalForce/ConstantAcceleration.cc
//---------------------------------Spheral++----------------------------------//
// ConstantAcceleration -- Impose a constant acceleration on a given set of
// nodes.
//
// Created by JMO, Mon Sep 27 23:01:15 PDT 2004
//----------------------------------------------------------------------------//
#include "ConstantAcceleration.hh"
#include "DataBase/DataBase.hh"
#include "DataBase/IncrementState.hh"
#include "Hydro/HydroFieldNames.hh"
#include "Field/FieldList.hh"
#include "Field/NodeIterators.hh"
#include "Utilities/DBC.hh"

using std::vector;
using std::string;
using std::pair;
using std::make_pair;
using std::cout;
using std::cerr;
using std::endl;
using std::min;
using std::max;
using std::abs;

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor.
//------------------------------------------------------------------------------
template<typename Dimension>
ConstantAcceleration<Dimension>::
ConstantAcceleration(const Vector a0,
                     const NodeList<Dimension>& nodeList,
                     const vector<int>& indices):
  GenericBodyForce<Dimension>(),
  ma0(a0),
  mNodeListPtr(&nodeList),
  mFlags("constant acceleration flags for " + nodeList.name(), nodeList, 0) {
  for (const int i: indices) {
    mFlags(i) = 1;
  }
}

//------------------------------------------------------------------------------
// Constructor for all nodes in NodeList.
//------------------------------------------------------------------------------
template<typename Dimension>
ConstantAcceleration<Dimension>::
ConstantAcceleration(const Vector a0,
                     const NodeList<Dimension>& nodeList):
  GenericBodyForce<Dimension>(),
  ma0(a0),
  mNodeListPtr(&nodeList),
  mFlags("constant acceleration flags for " + nodeList.name(), nodeList, 1) {
}

//------------------------------------------------------------------------------
// Destructor
//------------------------------------------------------------------------------
template<typename Dimension>
ConstantAcceleration<Dimension>::
~ConstantAcceleration() {
}

//------------------------------------------------------------------------------
// Apply the acceleration to the derivatives.
//------------------------------------------------------------------------------
template<typename Dimension>
void
ConstantAcceleration<Dimension>::
evaluateDerivatives(const typename Dimension::Scalar /*time*/,
                    const typename Dimension::Scalar /*dt*/,
                    const DataBase<Dimension>& /*dataBase*/,
                    const State<Dimension>& /*state*/,
                    StateDerivatives<Dimension>& derivs) const {

  // Get the acceleration state we'll be modifying.
  const typename State<Dimension>::KeyType key = State<Dimension>::buildFieldKey(IncrementState<Dimension, Field<Dimension, Vector> >::prefix() + HydroFieldNames::velocity,
                                                                                 mNodeListPtr->name());
  Field<Dimension, Vector>& DvDt = derivs.field(key, Vector::zero);

  // Increment the acceleration.
  const unsigned n = mNodeListPtr->numNodes();
  for (unsigned i = 0; i != n; ++i) {
    if (mFlags(i) == 1) {
      DvDt(i) += ma0;
    }
  }
}

//------------------------------------------------------------------------------
// Calculate the timestep constraint.
//------------------------------------------------------------------------------
template<typename Dimension>
typename ConstantAcceleration<Dimension>::TimeStepType
ConstantAcceleration<Dimension>::
dt(const DataBase<Dimension>& /*dataBase*/, 
   const State<Dimension>& /*state*/,
   const StateDerivatives<Dimension>& /*derivs*/,
   const typename Dimension::Scalar /*currentTime*/) const {

  return TimeStepType(FLT_MAX, "No vote.");
}

}

File: src/Pybind11Wraps/ExternalForce/ConstantAcceleration.py
#-------------------------------------------------------------------------------
# ConstantAcceleration base class
#-------------------------------------------------------------------------------
from PYB11Generator import *
from GenericBodyForce import *

@PYB11template("Dimension")
class ConstantAcceleration(GenericBodyForce):

    PYB11typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef typename %(Dimension)s::Vector Vector;
    typedef typename Physics<%(Dimension)s>::TimeStepType TimeStepType;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               a0 = "const Vector",
               nodeList = "const NodeList<%(Dimension)s>&",
               indices = "const std::vector<int>&"):
        "ConstantAcceleration constructor"

    def pyinit1(self,
                a0 = "const Vector",
                nodeList = "const NodeList<%(Dimension)s>&"):
        "ConstantAcceleration constructor"

    #...........................................................................
    # Virtual methods
    @PYB11virtual
    @PYB11const
    def evaluateDerivatives(self,
                            time = "const Scalar",
                            dt = "const Scalar",
                            dataBase = "const DataBase<%(Dimension)s>&",
                            state = "const State<%(Dimension)s>&",
                            derivs = "StateDerivatives<%(Dimension)s>&"):
        "Increment the derivatives."
        return "void"

    @PYB11virtual
    @PYB11const
    def dt(dataBase = "const DataBase<%(Dimension)s>&", 
           state = "const State<%(Dimension)s>&",
           derivs = "const StateDerivatives<%(Dimension)s>&",
           currentTime = "const Scalar"):
        "Vote on a time step."
        return "TimeStepType"

    @PYB11virtual
    @PYB11const
    def label(self):
        "It's useful to have labels for Physics packages.  We'll require this to have the same signature as the restart label."
        return "std::string"

    #...........................................................................
    # Properties
    a0 = PYB11property("Vector", "a0", doc="The fixed acceleration to apply")
    nodeList = PYB11property("const NodeList<%(Dimension)s>&", "nodeList", returnpolicy="reference_internal", doc="The NodeList this ConstantAcceleration applies to")
    flags = PYB11property("const Field<%(Dimension)s, int>&", "flags", returnpolicy="reference_internal", doc="The node flags")
